hazelcast/src/main/java/com/hazelcast/core/Pipelining.java

```diff
@@ -132,20 +125,57 @@ public Pipelining(int depth) {
      */
     public ICompletableFuture<E> add(ICompletableFuture<E> future) throws InterruptedException {
         checkNotNull(future, "future can't be null");
+        this.thread = Thread.currentThread();
 
-        semaphore.acquire();
+        down();
         futures.add(future);
         future.andThen(new ExecutionCallback<E>() {
             @Override
             public void onResponse(E response) {
-                semaphore.release();
+                up();
             }
 
             @Override
             public void onFailure(Throwable t) {
-                semaphore.release();
+                up();
             }
-        }, EXECUTOR);
+        }, CALLER_RUNS);
         return future;
     }
+
+    private void down() throws InterruptedException {
+        for (; ; ) {
+            int current = permits.get();
+            int update = current - 1;
+            if (!permits.compareAndSet(current, update)) {
+                // we failed to cas, so lets try again.
+                continue;
+            }
+
+            if (update == -1) {
+                // There are no more licences remaining, so we need to block
+                // Needs to be done in a loop to protect against spurious updates
```

Comment:
spurious wakeups.

Reply:
can't this be simplified into 
```java
while (permits.get() == -1) {
  LockSupport.park();
  if (Thread.interrupted()) {
    throw new InterruptedException();
  }
}
```
ie. without the outer `update == -1`

Reply:
Nice catch. I guess in the existing code we avoid a volatile read (`permits.get()`) in the fast path when there are permits?

Reply:
Doing a volatile read on this level is not going to make a big difference because there is so much code executed at a lower level.

Reply:
I just didn't look long enough at the code to see the cleanup.